sbi/src/tee_interrupt.rs

```diff
@@ -97,6 +97,55 @@ pub enum TeeInterruptFunction {
         /// a0 = physical address of interrupt file to be reclaimed
         imsic_addr: u64,
     },
+    /// Binds a TVM vCPU to the current physical CPU using the confidential guest interrupt files
+    /// specified in `imsic_mask`, restoring interrupt state from the vCPU's software interrupt
+    /// file if necessary. Note that `imsic_mask` is in the same format as the `hgeie` and `hgeip`
+    /// CSRs, that is bit N corresponds to guest interrupt file N-1 and bit 0 is always 0. The
+    /// number of bits set in `imsic_mask` must be equal to the number of interrupt files in the
+    /// vCPU's virtualized IMSIC (i.e. 1 + `guests_per_hart`). The vCPU must currently be unbound.
+    /// Upon completion,  the vCPU is eligible to be run on this CPU with `TvmCpuRun`.
+    ///
+    /// Returns 0 on succees.
+    ///
+    /// a6 = 4
+    TvmCpuBindImsic {
+        /// a0 = TVM ID
+        tvm_id: u64,
+        /// a1 = vCPU ID
+        vcpu_id: u64,
+        /// a2 = bit mask of interrupt files to be bound
+        imsic_mask: u64,
+    },
+    /// Begins the unbind process for the specified vCPU from its guest interrupt files. The
+    /// translations for the vCPU's virtualized IMSIC are invalidated, and a TLB flush sequence
+    /// for the TVM must be completed before calling `TvmCpuUnbindImsicEnd` to complete the unbind
+    /// process. Must be called on the physical CPU to which the vCPU is bound.
+    ///
+    /// Returns 0 on success.
+    ///
+    /// a6 = 5
+    TvmCpuUnbindImsicBegin {
```

Comment:
Does the unbind sequence look like this?

UnbindBegin
  TvmInitiateFence
    do needed IPIs
UnbindEnd

Did you consider automatically triggering the fence on UnbindBegin?

Reply:
> Does the unbind sequence look like this?
>
> UnbindBegin
TvmInitiateFence
do needed IPIs
UnbindEnd

Yes, that's right.

> Did you consider automatically triggering the fence on UnbindBegin?

I did. Two reasons I didn't go the implicit fence route:
1) It prevents the host from batching multiple invalidations in a single fence sequence
2) Since the fence sequence must be executed atomically, the host would need to hold a TVM-global lock around `unbind_begin()` + IPIs.

I was more concerned about (2) since `unbind_begin()` might be expensive. It has to: prepare the MRIF, unmap the IMSIC page, and issue IOMMU fence commands. 

Thoughts?

Reply:
Off the top of my head, the batch interface definitely seems like the optimal approach. Most IOMMUs that I know support batched invalidations for similar reasons.